crazy_rl/multi_agent/jax/catch/catch.py

```diff
@@ -0,0 +1,323 @@
+
+        return jdc.replace(
+            state,
+            rewards=jnp.any(state.truncations)
+            * (
+                # mean distance to the other agents
+                jnp.array(
+                    [
+                        jnp.sum(jnp.linalg.norm(state.agents_locations[agent] - state.agents_locations, axis=1))
+                        for agent in range(self.num_drones)
+                    ]
+                )
+                * 0.05
+                / (self.num_drones - 1)
+                # a maximum value minus the distance to the target
+                + 0.95 * (2 * self.size - jnp.linalg.norm(state.agents_locations - state.target_location, axis=1))
+            )
+            # negative reward if the drones crash
+            + jnp.any(state.terminations) * (1 - jnp.any(state.truncations)) * -10 * jnp.ones(self.num_drones),
+        )
+
+    @override
+    @partial(jit, static_argnums=(0,))
+    def _compute_terminated(self, state):
+        # collision with the ground and the target
+        terminated = jnp.logical_or(
+            state.agents_locations[:, 2] < 0.2, jnp.linalg.norm(state.agents_locations - state.target_location) < 0.2
+        )
+
+        for agent in range(self.num_drones):
+            distances = jnp.linalg.norm(state.agents_locations[agent] - state.agents_locations, axis=1)
+
+            # collision between two drones
+            terminated = terminated.at[agent].set(
+                jnp.logical_or(terminated[agent], jnp.any(jnp.logical_and(distances > 0.001, distances < 0.2)))
+            )
+
+        return jdc.replace(state, terminations=jnp.any(terminated) * jnp.ones(self.num_drones))
+
+    @override
+    @partial(jit, static_argnums=(0,))
+    def _compute_truncation(self, state):
+        return jdc.replace(state, truncations=(state.timestep == 100) * jnp.ones(self.num_drones))
+
+    @override
+    @partial(jit, static_argnums=(0,))
+    def reset(self, key):
+        """Resets the environment in initial state."""
+        state = State(
+            agents_locations=self._init_flying_pos,
+            timestep=0,
+            observations=jnp.array([]),
+            rewards=jnp.zeros(self.num_drones),
+            terminations=jnp.zeros(self.num_drones),
+            truncations=jnp.zeros(self.num_drones),
+            target_location=jnp.array([self._target_location]),
+        )
+        state = self._compute_obs(state)
+        return state
+
+    @override
+    @partial(jit, static_argnums=(0,))
+    def auto_reset(self, **state):
+        """Returns the State reinitialized if needed, else the actual State.
+
+        The values contained by State are passed in argument and used like a dictionary
+        because auto_reset is meant to be used by vmap and vmap doesn't accept objects.
+        """
+        done = jnp.any(state["truncations"]) + jnp.any(state["terminations"])
```

Comment:
I think there should be a jnp.max(...., 1) somewhere here to avoid truncation and terminations being true ... or a jnp.and(...)